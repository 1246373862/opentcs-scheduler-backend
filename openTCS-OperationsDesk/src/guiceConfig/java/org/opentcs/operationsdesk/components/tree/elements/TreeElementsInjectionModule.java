/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.components.tree.elements;

import com.google.inject.AbstractModule;
import com.google.inject.assistedinject.FactoryModuleBuilder;
import org.opentcs.guing.common.components.tree.elements.UserObjectFactory;
import org.opentcs.guing.common.components.tree.elements.VehicleUserObject;

/**
 * A Guice module for this package.
 */
public class TreeElementsInjectionModule
    extends AbstractModule {

  /**
   * Creates a new instance.
   */
  public TreeElementsInjectionModule() {
  }

  @Override
  protected void configure() {
    install(new FactoryModuleBuilder()
        .implement(VehicleUserObject.class, VehicleUserObjectOperating.class)
        .build(UserObjectFactory.class));
  }
}
