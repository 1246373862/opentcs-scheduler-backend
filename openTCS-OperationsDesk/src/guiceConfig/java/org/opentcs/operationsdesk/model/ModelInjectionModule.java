/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.model;

import com.google.inject.AbstractModule;
import org.opentcs.guing.common.model.SystemModel;

/**
 * A Guice module for the model package.
 */
public class ModelInjectionModule
    extends AbstractModule {

  /**
   * Creates a new instance.
   */
  public ModelInjectionModule() {
  }

  @Override
  protected void configure() {
    bind(SystemModel.class).to(CachedSystemModel.class);
  }

}
