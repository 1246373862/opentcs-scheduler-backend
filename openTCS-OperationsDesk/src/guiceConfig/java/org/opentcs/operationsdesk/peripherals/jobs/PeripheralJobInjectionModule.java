/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.peripherals.jobs;

import com.google.inject.AbstractModule;
import com.google.inject.assistedinject.FactoryModuleBuilder;
import javax.inject.Singleton;

/**
 * A Guice module for this package.
 */
public class PeripheralJobInjectionModule
    extends AbstractModule {

  /**
   * Creates a new instance.
   */
  public PeripheralJobInjectionModule() {
  }

  @Override
  protected void configure() {
    install(new FactoryModuleBuilder().build(PeripheralJobViewFactory.class));
    bind(PeripheralJobsContainer.class).in(Singleton.class);
  }
}
