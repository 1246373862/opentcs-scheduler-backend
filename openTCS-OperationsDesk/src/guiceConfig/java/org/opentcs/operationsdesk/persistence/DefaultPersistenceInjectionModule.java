/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.persistence;

import com.google.inject.AbstractModule;
import com.google.inject.Singleton;
import org.opentcs.guing.common.persistence.ModelFilePersistor;
import org.opentcs.guing.common.persistence.ModelManager;
import org.opentcs.guing.common.persistence.OpenTCSModelManager;
import org.opentcs.guing.common.persistence.unified.UnifiedModelPersistor;

/**
 * Default bindings for model readers and persistors.
 */
public class DefaultPersistenceInjectionModule
    extends AbstractModule {

  /**
   * Creates a new instance.
   */
  public DefaultPersistenceInjectionModule() {
  }

  @Override
  protected void configure() {
    bind(ModelManager.class).to(OpenTCSModelManager.class).in(Singleton.class);

    bind(ModelFilePersistor.class).to(UnifiedModelPersistor.class);
  }

}
