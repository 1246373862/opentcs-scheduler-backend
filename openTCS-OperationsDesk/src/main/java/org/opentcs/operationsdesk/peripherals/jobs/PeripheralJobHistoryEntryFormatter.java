/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.peripherals.jobs;

import static java.util.Objects.requireNonNull;
import java.util.Optional;
import org.opentcs.components.plantoverview.ObjectHistoryEntryFormatter;
import org.opentcs.data.ObjectHistory;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.data.peripherals.PeripheralJobHistoryCodes;
import org.opentcs.operationsdesk.util.I18nPlantOverviewOperating;
import org.opentcs.thirdparty.guing.common.jhotdraw.util.ResourceBundleUtil;

/**
 * A formatter for history events/entries related to {@link PeripheralJob}s.
 */
public class PeripheralJobHistoryEntryFormatter
    implements ObjectHistoryEntryFormatter {

  /**
   * A bundle providing localized strings.
   */
  private final ResourceBundleUtil bundle
      = ResourceBundleUtil.getBundle(I18nPlantOverviewOperating.PJDETAIL_PATH);

  /**
   * Creates a new instance.
   */
  public PeripheralJobHistoryEntryFormatter() {
  }

  @Override
  public Optional<String> apply(ObjectHistory.Entry entry) {
    requireNonNull(entry, "entry");

    switch (entry.getEventCode()) {
      case PeripheralJobHistoryCodes.JOB_CREATED:
        return Optional.of(
            bundle.getString("peripheralJobHistoryEntryFormatter.code_jobCreated.text")
        );
      case PeripheralJobHistoryCodes.JOB_REACHED_FINAL_STATE:
        return Optional.of(
            bundle.getString("peripheralJobHistoryEntryFormatter.code_jobReachedFinalState.text")
        );

      default:
        return Optional.empty();
    }
  }

}
