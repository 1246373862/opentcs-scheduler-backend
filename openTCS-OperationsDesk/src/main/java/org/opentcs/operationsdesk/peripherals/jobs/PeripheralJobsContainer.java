/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.peripherals.jobs;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import static java.util.Objects.requireNonNull;
import java.util.Set;
import javax.inject.Inject;
import org.opentcs.access.KernelRuntimeException;
import org.opentcs.access.SharedKernelServicePortal;
import org.opentcs.access.SharedKernelServicePortalProvider;
import org.opentcs.components.Lifecycle;
import org.opentcs.customizations.ApplicationEventBus;
import org.opentcs.data.TCSObjectEvent;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.guing.common.event.OperationModeChangeEvent;
import org.opentcs.guing.common.event.SystemModelTransitionEvent;
import org.opentcs.operationsdesk.event.KernelStateChangeEvent;
import org.opentcs.util.event.EventBus;
import org.opentcs.util.event.EventHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Maintains a set of all peripheral jobs existing on the kernel side.
 */
public class PeripheralJobsContainer
    implements EventHandler,
               Lifecycle {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(PeripheralJobsContainer.class);
  /**
   * Where we get events from.
   */
  private final EventBus eventBus;
  /**
   * Provides access to a portal.
   */
  private final SharedKernelServicePortalProvider portalProvider;
  /**
   * The peripheral jobs.
   */
  private final Map<String, PeripheralJob> peripheralJobs = new HashMap<>();
  /**
   * This container's listeners.
   */
  private final Set<PeripheralJobsContainerListener> listeners = new HashSet<>();
  /**
   * Whether this component is initialized.
   */
  private boolean initialized;

  /**
   * Creates a new instance.
   *
   * @param eventBus Where this instance subscribes for events.
   * @param portalProvider Provides a access to a portal.
   */
  @Inject
  public PeripheralJobsContainer(@ApplicationEventBus EventBus eventBus,
                                 SharedKernelServicePortalProvider portalProvider) {
    this.eventBus = requireNonNull(eventBus, "eventBus");
    this.portalProvider = requireNonNull(portalProvider, "portalProvider");
  }

  @Override
  public boolean isInitialized() {
    return initialized;
  }

  @Override
  public void initialize() {
    if (isInitialized()) {
      return;
    }

    eventBus.subscribe(this);

    initialized = true;
  }

  @Override
  public void terminate() {
    if (!isInitialized()) {
      return;
    }

    eventBus.unsubscribe(this);

    initialized = false;
  }

  @Override
  public void onEvent(Object event) {
    if (event instanceof TCSObjectEvent) {
      handleObjectEvent((TCSObjectEvent) event);
    }
    else if (event instanceof OperationModeChangeEvent) {
      initJobs();
    }
    else if (event instanceof SystemModelTransitionEvent) {
      initJobs();
    }
    else if (event instanceof KernelStateChangeEvent) {
      initJobs();
    }
  }

  public void addListener(PeripheralJobsContainerListener listener) {
    listeners.add(listener);
  }

  public void removeListener(PeripheralJobsContainerListener listener) {
    listeners.remove(listener);
  }

  public Collection<PeripheralJob> getPeripheralJobs() {
    return peripheralJobs.values();
  }

  private void initJobs() {
    setPeripheralJobs(fetchJobsIfOnline());
    listeners.forEach(listener -> listener.containerInitialized(peripheralJobs.values()));
  }

  private void handleObjectEvent(TCSObjectEvent evt) {
    if (evt.getCurrentOrPreviousObjectState() instanceof PeripheralJob) {
      switch (evt.getType()) {
        case OBJECT_CREATED:
          peripheralJobAdded((PeripheralJob) evt.getCurrentOrPreviousObjectState());
          break;
        case OBJECT_MODIFIED:
          peripheralJobChanged((PeripheralJob) evt.getCurrentOrPreviousObjectState());
          break;
        case OBJECT_REMOVED:
          peripheralJobRemoved((PeripheralJob) evt.getCurrentOrPreviousObjectState());
          break;
        default:
          LOG.warn("Unhandled event type: {}", evt.getType());
      }
    }
  }

  private void peripheralJobAdded(PeripheralJob job) {
    peripheralJobs.put(job.getName(), job);
    listeners.forEach(listener -> listener.peripheralJobAdded(job));
  }

  private void peripheralJobChanged(PeripheralJob job) {
    peripheralJobs.put(job.getName(), job);
    listeners.forEach(listener -> listener.peripheralJobUpdated(job));
  }

  private void peripheralJobRemoved(PeripheralJob job) {
    peripheralJobs.remove(job.getName());
    listeners.forEach(listener -> listener.peripheralJobRemoved(job));
  }

  private void setPeripheralJobs(Set<PeripheralJob> newJobs) {
    peripheralJobs.clear();
    for (PeripheralJob job : newJobs) {
      peripheralJobs.put(job.getName(), job);
    }
  }

  private Set<PeripheralJob> fetchJobsIfOnline() {
    if (portalProvider.portalShared()) {
      try (SharedKernelServicePortal sharedPortal = portalProvider.register()) {
        return sharedPortal.getPortal().getPeripheralJobService().fetchObjects(PeripheralJob.class);
      }
      catch (KernelRuntimeException exc) {
        LOG.warn("Exception fetching peripheral jobs", exc);
      }
    }

    return new HashSet<>();
  }
}
