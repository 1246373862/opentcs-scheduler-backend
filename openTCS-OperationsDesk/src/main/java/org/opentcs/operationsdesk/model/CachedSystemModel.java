/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.model;

import java.util.HashMap;
import java.util.Map;
import javax.inject.Inject;
import org.opentcs.guing.base.model.ModelComponent;
import org.opentcs.guing.base.model.elements.BlockModel;
import org.opentcs.guing.base.model.elements.LocationModel;
import org.opentcs.guing.base.model.elements.LocationTypeModel;
import org.opentcs.guing.base.model.elements.PathModel;
import org.opentcs.guing.base.model.elements.PointModel;
import org.opentcs.guing.base.model.elements.VehicleModel;
import org.opentcs.guing.common.model.StandardSystemModel;
import org.opentcs.guing.common.util.ModelComponentFactory;

/**
 * Extends the standard system model with a cache of model components
 * to provide a more efficient lookup of components by name.
 */
public class CachedSystemModel
    extends StandardSystemModel {

  /**
   * A map from component name to the actual component.
   */
  private final Map<String, ModelComponent> components = new HashMap<>();

  @Inject
  public CachedSystemModel(ModelComponentFactory modelComponentFactory) {
    super(modelComponentFactory);
  }

  @Override
  public void onRestorationComplete() {
    components.clear();
    for (ModelComponent component : getAll()) {
      components.put(component.getName(), component);
    }
  }

  @Override
  public BlockModel getBlockModel(String name) {
    if (components.isEmpty()) {
      return super.getBlockModel(name);
    }

    ModelComponent block = components.get(name);
    if (block instanceof BlockModel) {
      return (BlockModel) block;
    }
    return null;
  }

  @Override
  public LocationModel getLocationModel(String name) {
    if (components.isEmpty()) {
      return super.getLocationModel(name);
    }

    ModelComponent location = components.get(name);
    if (location instanceof LocationModel) {
      return (LocationModel) location;
    }
    return null;
  }

  @Override
  public LocationTypeModel getLocationTypeModel(String name) {
    if (components.isEmpty()) {
      return super.getLocationTypeModel(name);
    }

    ModelComponent locationType = components.get(name);
    if (locationType instanceof LocationTypeModel) {
      return (LocationTypeModel) locationType;
    }
    return null;
  }

  @Override
  public ModelComponent getModelComponent(String name) {
    if (components.isEmpty()) {
      return super.getModelComponent(name);
    }
    return components.get(name);
  }

  @Override
  public PathModel getPathModel(String name) {
    if (components.isEmpty()) {
      return super.getPathModel(name);
    }

    ModelComponent path = components.get(name);
    if (path instanceof PathModel) {
      return (PathModel) path;
    }
    return null;
  }

  @Override
  public PointModel getPointModel(String name) {
    if (components.isEmpty()) {
      return super.getPointModel(name);
    }

    ModelComponent point = components.get(name);
    if (point instanceof PointModel) {
      return (PointModel) point;
    }
    return null;
  }

  @Override
  public VehicleModel getVehicleModel(String name) {
    if (components.isEmpty()) {
      return super.getVehicleModel(name);
    }

    ModelComponent vehicle = components.get(name);
    if (vehicle instanceof VehicleModel) {
      return (VehicleModel) vehicle;
    }
    return null;
  }

}
