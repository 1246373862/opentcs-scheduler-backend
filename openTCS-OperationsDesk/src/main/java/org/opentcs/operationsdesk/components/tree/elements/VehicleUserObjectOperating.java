/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.components.tree.elements;

import com.google.inject.assistedinject.Assisted;
import static java.util.Objects.requireNonNull;
import javax.inject.Inject;
import javax.swing.JPopupMenu;
import org.opentcs.guing.base.model.elements.VehicleModel;
import org.opentcs.guing.common.application.GuiManager;
import org.opentcs.guing.common.components.tree.elements.VehicleUserObject;
import org.opentcs.guing.common.persistence.ModelManager;
import org.opentcs.operationsdesk.application.menus.MenuFactory;

/**
 * A Vehicle in the tree view.
 */
public class VehicleUserObjectOperating
    extends VehicleUserObject {

  /**
   * A factory for popup menus.
   */
  private final MenuFactory menuFactory;

  /**
   * Creates a new instance.
   *
   * @param model The corresponding vehicle object.
   * @param guiManager The gui manager.
   * @param modelManager Provides the current system model.
   * @param menuFactory A factory for popup menus.
   */
  @Inject
  public VehicleUserObjectOperating(@Assisted VehicleModel model,
                                    GuiManager guiManager,
                                    ModelManager modelManager,
                                    MenuFactory menuFactory) {
    super(model, guiManager, modelManager);
    this.menuFactory = requireNonNull(menuFactory, "menuFactory");
  }

  @Override
  public JPopupMenu getPopupMenu() {
    return menuFactory.createVehiclePopupMenu(selectedVehicles);
  }
}
