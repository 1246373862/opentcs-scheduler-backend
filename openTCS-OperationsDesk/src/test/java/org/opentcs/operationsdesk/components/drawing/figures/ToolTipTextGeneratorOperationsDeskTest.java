/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.operationsdesk.components.drawing.figures;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Locale;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import org.opentcs.data.model.Location;
import org.opentcs.data.model.LocationType;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.data.peripherals.PeripheralOperation;
import org.opentcs.guing.base.AllocationState;
import org.opentcs.guing.base.model.elements.LocationModel;
import org.opentcs.guing.base.model.elements.PathModel;
import org.opentcs.guing.base.model.elements.PointModel;
import org.opentcs.guing.base.model.elements.VehicleModel;
import org.opentcs.guing.common.model.SystemModel;
import org.opentcs.guing.common.persistence.ModelManager;
import org.opentcs.operationsdesk.peripherals.jobs.PeripheralJobsContainer;

/**
 * Unit tests for {@link ToolTipTextGeneratorOperationsDesk}.
 */
public class ToolTipTextGeneratorOperationsDeskTest {

  private ToolTipTextGeneratorOperationsDesk toolTipTextGenerator;

  private PeripheralJobsContainer peripheralJobsContainer;

  private Vehicle vehicle;
  private VehicleModel vehicleModel;
  private Location location;
  private LocationType locationType;

  @BeforeEach
  public void setup() {
    Locale.setDefault(Locale.forLanguageTag("en"));

    SystemModel systemModel = mock(SystemModel.class);
    when(systemModel.getBlockModels()).thenReturn(new ArrayList<>());

    ModelManager modelManager = mock(ModelManager.class);
    when(modelManager.getModel()).thenReturn(systemModel);

    peripheralJobsContainer = mock(PeripheralJobsContainer.class);
    toolTipTextGenerator = new ToolTipTextGeneratorOperationsDesk(
        modelManager,
        peripheralJobsContainer
    );

    vehicle = new Vehicle("Vehicle-001");
    vehicleModel = new VehicleModel();
    vehicleModel.setVehicle(vehicle);
    vehicleModel.setName(vehicle.getName());

    locationType = new LocationType("Loc-Type-001");
    location = new Location("Location-001", locationType.getReference());
  }

  @Test
  public void listNoPeripheralJobs() {
    Approvals.verify(toolTipTextGenerator.getToolTipText(vehicleModel));
  }

  @Test
  public void listOnlyPeripheralJobsRelatedToVehicle() {
    PeripheralJob relatedJob = new PeripheralJob(
        "Job-name-001",
        "ReservationToken-01",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference());
    PeripheralJob unrelatedJob = new PeripheralJob(
        "Job-name-002",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(new Vehicle("Other Vehicle").getReference());
    PeripheralJob jobWithNoRelatedVehicle = new PeripheralJob(
        "Job-name-003",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    );

    when(peripheralJobsContainer.getPeripheralJobs())
        .thenReturn(Arrays.asList(relatedJob, unrelatedJob, jobWithNoRelatedVehicle));

    Approvals.verify(toolTipTextGenerator.getToolTipText(vehicleModel));
  }

  @Test
  public void listOnlyPeripheralJobsWithCompletionRequired() {
    PeripheralJob completionRequiredJob = new PeripheralJob(
        "Job-name-001",
        "ReservationToken-01",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference());
    PeripheralJob completionNotequiredJob = new PeripheralJob(
        "Job-name-002",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            false
        )
    ).withRelatedVehicle(vehicle.getReference());

    when(peripheralJobsContainer.getPeripheralJobs())
        .thenReturn(Arrays.asList(completionRequiredJob, completionNotequiredJob));

    Approvals.verify(toolTipTextGenerator.getToolTipText(vehicleModel));
  }

  @Test
  public void listOnlyPeripheralJobsInNonFinalState() {
    PeripheralJob job1 = new PeripheralJob(
        "Approved-job-01",
        "ReservationToken-01",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference())
        .withState(PeripheralJob.State.TO_BE_PROCESSED);
    PeripheralJob job2 = new PeripheralJob(
        "Approved-job-02",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference())
        .withState(PeripheralJob.State.BEING_PROCESSED);
    PeripheralJob job3 = new PeripheralJob(
        "Rejected-job-01",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference())
        .withState(PeripheralJob.State.FAILED);
    PeripheralJob job4 = new PeripheralJob(
        "Rejected-job-02",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference())
        .withState(PeripheralJob.State.FINISHED);

    when(peripheralJobsContainer.getPeripheralJobs())
        .thenReturn(Arrays.asList(job1, job2, job3, job4));

    Approvals.verify(toolTipTextGenerator.getToolTipText(vehicleModel));
  }

  @Test
  public void sortPeripheralJobsByCreationTime() {
    PeripheralJob firstJob = new PeripheralJob(
        "This one first",
        "ReservationToken-01",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference())
        .withCreationTime(Instant.ofEpochSecond(10));
    PeripheralJob secondJob = new PeripheralJob(
        "This one second",
        "ReservationToken-02",
        new PeripheralOperation(
            location.getReference(),
            "PeripheralOperation",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    ).withRelatedVehicle(vehicle.getReference())
        .withCreationTime(Instant.ofEpochSecond(999));

    when(peripheralJobsContainer.getPeripheralJobs())
        .thenReturn(Arrays.asList(secondJob, firstJob));

    Approvals.verify(toolTipTextGenerator.getToolTipText(vehicleModel));
  }

  @Test
  public void addAllocatingVehicleForPoints() {
    Vehicle vehicle2 = new Vehicle("Vehicle-002");
    VehicleModel vehicleModel2 = new VehicleModel();
    vehicleModel2.setVehicle(vehicle2);
    vehicleModel2.setName(vehicle2.getName());

    PointModel point = new PointModel();
    point.setName("Point-0001");
    point.updateAllocationState(vehicleModel, AllocationState.ALLOCATED);
    point.updateAllocationState(vehicleModel2, AllocationState.CLAIMED);

    Approvals.verify(toolTipTextGenerator.getToolTipText(point));
  }

  @Test
  public void addAllocatingVehicleForPaths() {
    Vehicle vehicle2 = new Vehicle("Vehicle-002");
    VehicleModel vehicleModel2 = new VehicleModel();
    vehicleModel2.setVehicle(vehicle2);
    vehicleModel2.setName(vehicle2.getName());

    PathModel path = new PathModel();
    path.setName("Path-0001");
    path.updateAllocationState(vehicleModel, AllocationState.ALLOCATED);
    path.updateAllocationState(vehicleModel2, AllocationState.CLAIMED);

    Approvals.verify(toolTipTextGenerator.getToolTipText(path));
  }

  @Test
  public void addAllocatingVehicleForLocations() {
    Vehicle vehicle2 = new Vehicle("Vehicle-002");
    VehicleModel vehicleModel2 = new VehicleModel();
    vehicleModel2.setVehicle(vehicle2);
    vehicleModel2.setName(vehicle2.getName());

    LocationModel location = new LocationModel();
    location.setName("Location-0001");
    location.updateAllocationState(vehicleModel, AllocationState.ALLOCATED);
    location.updateAllocationState(vehicleModel2, AllocationState.CLAIMED);

    Approvals.verify(toolTipTextGenerator.getToolTipText(location));
  }

}
