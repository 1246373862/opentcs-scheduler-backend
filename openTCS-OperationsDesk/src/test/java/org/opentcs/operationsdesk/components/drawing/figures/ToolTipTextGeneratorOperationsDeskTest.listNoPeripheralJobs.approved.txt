<html>
Vehicle <b>Vehicle-001</b>
<hr>
Current point: ?
<br>
Next point: ?
<br>
State: <font color=orange>UNKNOWN</font>
<br>
Processing state: IDLE
<br>
Integration level: TO_BE_RESPECTED
<br>
Current energy level: <font color=green>0%</font>
<hr>
Peripheral operations that need to be waited for:
<ul>
</ul>
</html>
